File: echo/test/echo_test.go
package lib

import (
	"bytes"
	"io"
	"net/http"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

func TestEchoConcurrent(t *testing.T) {
	// Set the number of concurrent requests
	concurrency := 30000

	// Atomic counter
	var success int32

	// Create a reusable HTTP client
	client := &http.Client{}

	// Launch concurrent requests
	var wg sync.WaitGroup
	wg.Add(concurrency)
	start := time.Now() // Record start time
	for i := 0; i < concurrency; i++ {
		go func() {
			defer wg.Done()
			// Generate a request
			req, err := http.NewRequest("POST", "http://localhost:8080/echo", bytes.NewBuffer([]byte("Hello, World!")))
			if err != nil {
				t.Logf("Error: %v", err) // Log error
				return
			}

			// Send the request
			resp, err := client.Do(req)
			if err != nil {
				t.Logf("Error: %v", err) // Log error
				return
			}

			// Read and discard the response body
			_, err = io.ReadAll(resp.Body)
			resp.Body.Close()
			if err != nil {
				t.Logf("Error: %v", err) // Log error
				return
			}

			// Update success count using atomic operation
			atomic.AddInt32(&success, 1)
		}()
	}

	// Wait for all concurrent requests to complete
	wg.Wait()
	elapsed := time.Since(start)           // Calculate test duration
	t.Logf("Total time: %v", elapsed)      // Log total time
	t.Logf("Success request: %d", success) // Log successful requests count
}

File: echo/test/echo_test.js
const axios = require('axios');
const { performance } = require('perf_hooks');

const concurrency = 30000; // 並發請求數
let success = 0; // 成功請求計數
let failed = 0; // 失敗請求計數

const start = performance.now(); // 記錄開始時間

const requests = Array.from({ length: concurrency }, () => {
  return axios.post('http://localhost:3000/', 'Hello, World!')
    .then(() => {
      success++;
    })
    .catch(() => {
      failed++;
    });
});

Promise.all(requests)
  .then(() => {
    const end = performance.now(); // 記錄結束時間
    const elapsed = end - start; // 計算執行時間
    console.log(`Total time: ${elapsed}ms`);
    console.log(`Success requests: ${success}`);
    console.log(`Failed requests: ${failed}`);
  })
  .catch(error => {
    console.error('An error occurred:', error);
  });
